src/stan/services/util/run_adaptive_sampler.hpp

```diff
@@ -87,6 +88,83 @@ void run_adaptive_sampler(Sampler& sampler, Model& model,
                           / 1000.0;
   writer.write_timing(warm_delta_t, sample_delta_t);
 }
+
+template <class Sampler, class Model, class RNG, typename SampT,
+          typename DiagnoseT>
+void run_adaptive_sampler(std::vector<Sampler>& samplers, Model& model,
+                          std::vector<std::vector<double>>& cont_vectors,
+                          int num_warmup, int num_samples, int num_thin,
+                          int refresh, bool save_warmup, std::vector<RNG>& rngs,
+                          callbacks::interrupt& interrupt,
+                          callbacks::logger& logger,
+                          std::vector<SampT>& sample_writers,
+                          std::vector<DiagnoseT>& diagnostic_writers,
+                          size_t n_chain) {
```

Comment:
So the model, interrupt, and logger bits need to be threadsafe.

Is every function in the model (other than constructors/destructors) thread safe? I think this threading stuff will mess up the profiles, so we gotta figure out what to do there.

For the interrupts and loggers, I see that the implementations in the test had to change but shouldn't there be changes in the main source too? Or were these already threadsafe or something?

I see that interrupt is for the interfaces to capture keystrokes. Are the interrupts signals generated at the interfaces and transmitted in here -- or are they signals generated in the samplers and transmitted back to the interfaces?

Reply:
So the logger goes to `std::cout` which is threadsafe (see [here](https://en.cppreference.com/w/cpp/io/cout))

>> Unless sync_with_stdio(false) has been issued, it is safe to concurrently access these objects from multiple threads for both formatted and unformatted output. 

The interrupts, at least in cmdstan, are actually just the base impl `stan::callbacks::interrupt` which doesn't actually do anything. I'm looking at the [R interrupt](https://rdrr.io/github/HenrikBengtsson/RNativeAPI/man/R_CheckUserInterrupt.html) and it seems like that would be threadsafe. 

> Is every function in the model (other than constructors/destructors) thread safe?

Yep! I've looked the model class over. The model is essentially immutable after construction so it's safe to pass around.

>  I think this threading stuff will mess up the profiles, so we gotta figure out what to do there.

Oooh hm idk how this will effect profiling. @rok-cesnovar looking at the profiling gen code

```
      {
        profile<local_scalar_t__> profile__("testerrr", const_cast<profile_map&>(profiles__));
        
        current_statement__ = 3;
        assign(transf_testp, stan::math::exp(testp),
          "assigning variable transf_testp");
      }
```

I think we might need a mutex in the profile constructor here

```cpp
  profile(std::string name, profile_map& profiles)
      : key_({name, std::this_thread::get_id()}) {
    {
      std::lock_guard<std::mutex> map_lock(profile_mutex);
     // idt search is thread safe
      profile_map::iterator p = profiles.find(key_);
      if (p == profiles.end()) {
        // idt insertion is thread safe either
        profiles[key_] = profile_info();
    }
    // rest business as usual
    }
```

> I see that interrupt is for the interfaces to capture keystrokes. Are the interrupts signals generated at the interfaces and transmitted in here -- or are they signals generated in the samplers and transmitted back to the interfaces?

They are called in [`generate_transitions`](https://github.com/stan-dev/stan/blob/develop/src/stan/services/util/generate_transitions.hpp#L49) which is for the keystroke thing for R / Python / higher level systems that need to tell the C++ to stop. It's not used in cmdstan since you can just call ctrl+c on a binary to tell the system to stop whatever program is running atm.

Reply:
Is it an option to run the profiler in only one thread or avoid the need for a mutex by using multiple model copies? We do not want a mutex unless we have to.

Reply:
I think we want separate profiles for each thread if we want profiling to act like it currently does (a separate profile for each chain). Is Stan data actually copied into the model? Or is it references? I guess if it's references then we could actually have lots of models? But maybe that is undesirable.

Reply:
Do we need a mutex? The thread id is a key in the storage map and each thread is thus guaranteed to always access only its "own" profiles. `std::map` is supposed to be fine in that case: https://stackoverflow.com/questions/15067160/stdmap-thread-safety and other resources if you google "std::map thread safety".

We could switch to using tbb::concurent_hash_map though I think that is unecessary.

Reply:
The profiler itself runs in one thread, the problem is that in the above constructor the `profile_map` is a global `std::map` and idt there's any reasonable way to get around that. 

imo idt a mutex is that bad here. I think we can actually put it in the `if (p == profiles.end())` since `find` should be threadsafe since it's searching for a key that also has the thread_id in it. But for insertion with `profiles[key_]` we do not want to be adding keys simultaneously as we could end up in a situation where the map reallocs memory


One solution here is to prefill the profile_map keys. Since we know each key is decided by a unique name and thread combination no two threads can cause a race condition because they can write to different parts of the map. So we only need to worry about the map reallocing for new keys. for an example model like  


```stan
model {
  profile("ayyy") {
    theta ~ beta(1,1);  // uniform prior on interval 0,1
  }
  profile("yooo") {
    y ~ bernoulli(theta);    
  }
}
```

We can have a function in the model like 

```
void init_profile_map(){
   const auto max_threads = get_max_threads();
   for (auto&& id : std::array<const char*, 2>{"ayyy", "yooo")) {
     for (int i = 0; i < max_threads; ++i) {
       profile_map[{name, std::this_thread::get_id()}] = empty_profile_info();
     }
   } 
}
```

Or something like that to fill up the keys before we do any other insertions. Then we don't need to worry about data reallocation since all the keys are there.

But still I think a mutex here is not that bad. It's short lived and only causes locks whenever we are doing profiling (which imo is a usage pattern that is not for performance anyway). I think what I would like is to just put a mutex in the if for now and then later we can do the lock free pattern

Reply:
Oh shoot I typed that out before Rok posted his reply. Yeah from there it looks like it is safe to insert since each key is distinct for each thread. I'll look into this more to make sure it's fine. If I can't sort out anything conclusive then I think moving to a ` tbb::concurent_hash_map` is a reasonable solution

Reply:
Global variables are problematic for obvious reasons. The TBB data structures should be a good alternative... How often is this code being run over? I mean, is it performance critical?

Reply:
Insert to a map happens the first time a profile is registered. So once in a lifetime of a model. After that its just read access to get the reference to profile info. And those are separated for separate threads, no race conditions there.

If locking the insert in a mutex is something that would make us sleep more calmly, thats not a big price to pay. It does seem its not needed.

Reply:
Yeah idt it's needed. Let's not worry about this for now and we should notice it with a pretty simple test once the cmdstan stuff is up